### PR TITLE
samples: threads: Rewrite misleading sample

### PR DESCRIPTION
**Pull request by SebastianBoe**
The basic/threads sample is printing a log that looks like this:

Toggle USR0 LED: Counter = 0
Toggle USR1 LED: Counter = 0
Toggle USR0 LED: Counter = 1
Toggle USR0 LED: Counter = 2
Toggle USR0 LED: Counter = 3
Toggle USR0 LED: Counter = 4

From that log you would think that it was logging when leds were
blinking, but actually the led-threads might have crashed, and it
would just continue logging anyway. All it's doing is executing
printk's at roughly the same frequency as the LED's are blinking.

This patch rewrites the sample to use a FIFO so that the printk's only
trigger if the LEDs are actually blinking.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

### CODE REVIEW

File: samples/basic/threads/src/main.c [hunk 1]
```diff
@@ -48,23 +59,20 @@ void blink2(void)
 	gpio_pin_configure(gpiod, LED1, GPIO_DIR_OUT);
 	while (1) {
 		gpio_pin_write(gpiod, LED1, cnt % 2);
+
+		struct printk_data_t tx_data = { .led = 1, .cnt = cnt };
```
Comment (cvinayak) [inline samples/basic/threads/src/main.c]: This should be from a mem pool, else the fifo "next" will start pointing to itself without the k_sleep ?

**Review by cvinayak — DISMISSED** [review]

File: samples/basic/threads/src/main.c (hunk 1 above)
Comment (SebastianBoe) [reply samples/basic/threads/src/main.c]: Sorry, I don't know what a mem pool is.

I think the code is valid, I used this documentation to write it:
http://docs.zephyrproject.org/kernel/data_passing/fifos.html#reading-from-a-fifo

Comment (SebastianBoe) [reply samples/basic/threads/src/main.c]: Can you elaborate on this:

" the fifo "next" will start pointing to itself without the k_sleep"

?

Comment (SebastianBoe) [reply samples/basic/threads/src/main.c]: Ah, I am sending the fifo a stack address, my bad, will fix.

Comment (SebastianBoe) [reply samples/basic/threads/src/main.c]: Fixed.

Comment (SebastianBoe) [reply samples/basic/threads/src/main.c]: Good catch, thank you for reviewing.

File: samples/basic/threads/README.rst [hunk 2]
```diff
@@ -8,11 +8,15 @@ Overview
...
-The example works by spawning  three threads. The first two threads control a
+The example works by spawning three threads. The first two threads control a
 separate LED. Both of these LEDs (USR1 and USR2) have their individual loop
 control and timing logic defined by separate functions.
 
-The third thread, uart_out(), sends out messages on the UART using printk().
+After a thread has toggled an LED it will push information about what
+thread toggled the LED and how many times it has done into a FIFO.
+
+The third thread, uart_out(), will printk the information that comes
```
Comment (dbkinder) [inline samples/basic/threads/README.rst]: How about:

    The third thread, ``uart_out()``, uses printk (over the UART) to display
    the information that comes through the FIFO.

**Review by dbkinder — DISMISSED** [review]
suggested edits

File: samples/basic/threads/README.rst (hunk 2 above)
Comment (SebastianBoe) [reply samples/basic/threads/README.rst]: Done.

File: samples/basic/threads/README.rst [hunk 3]
```diff
@@ -8,11 +8,15 @@ Overview
 
 This example demonstrates spawning of multiple threads using K_THREAD_DEFINE.
 
-The example works by spawning  three threads. The first two threads control a
+The example works by spawning three threads. The first two threads control a
 separate LED. Both of these LEDs (USR1 and USR2) have their individual loop
 control and timing logic defined by separate functions.
 
-The third thread, uart_out(), sends out messages on the UART using printk().
+After a thread has toggled an LED it will push information about what
```
[the remark this answers is not in the document]
Comment (SebastianBoe) [reply samples/basic/threads/README.rst]: Done.

**Comment by pfalcon** [comment]
Why not add the original author to the review? @ric96 : Ping. (The sample comes from https://github.com/zephyrproject-rtos/zephyr/pull/4489 & https://github.com/zephyrproject-rtos/zephyr/pull/4573).

**Comment by SebastianBoe** [comment]
> Why not add the original author to the review?
I was not able to find him based on the git blame.

Thanks.

**Review by dbkinder — APPROVED** [review]
thanks.

**Comment by ric96** [comment]
why not just use  gpio_pin_read in the uart function?

**Comment by SebastianBoe** [comment]
Not sure what you mean, then we would have to introduce the ISR concept to wake up the UART
thread wouldn't we? I would prefer not to do that in the sample named basic/threads.

**Comment by ric96** [comment]
@SebastianBoe yes, sorry about that. Your fix looks fine.
+1

**Pull request merged**